### PR TITLE
Do not require all extras for SalesforceHook

### LINKED ISSUE #19506: Salesforce connections should not require all extras
### Apache Airflow Provider(s)

salesforce

### Versions of Apache Airflow Providers

apache-airflow-providers-salesforce==1!3.2.0

### Apache Airflow version

2.1.4

### Operating System

Debian GNU/Linux 10 (buster)

### Deployment

Astronomer

### Deployment details

_No response_

### What happened

Right now the `SalesforceHook` is requiring all of the extras for a Salesforce connection type in Airflow.

I am assuming this issue was never brought up before because users of this hook have been using the Airflow UI to make connections (which presents all extra fields), however with things such as Secrets Backends, it should be possible to set up a Salesforce connection URI without having to explicitly provide all of the extras.

The issue is the hook's author designed the hook's `get_conn` to pass in every extra value using an improper/invalid method of defaulting to None by using `or None` but still referencing the extras key directly. I tested this in as low as Python 2.7 and as high as Python 3.7 and in both versions if any one of these extras are not provided you will get a `KeyError`

https://github.com/apache/airflow/blob/e9a72a4e95e6d23bae010ad92499cd7b06d50037/airflow/providers/salesforce/hooks/salesforce.py#L137-L149

### What you expected to happen

Any extras value not provided should just default to None (or to `api.DEFAULT_API_VERSION` for the "version" extra). It should be possible to set up a Salesforce connection URI using a secrets backend without having to explicitly provide all of the extras.

### How to reproduce

Set up a secrets backend (such as via environment variable) and pass in the minimum connection values needed for "Password" connection type:

_(this is an example, no real passwords shown)_
`export AIRFLOW_CONN_SALESFORCE_DEFAULT='http://your_username:your_password@https%3A%2F%2Fyour_host.lightning.force.com?extra__salesforce__security_token=your_token'`

It will error with `KeyError: 'extra__salesforce__domain'` and keep resulting in key errors for each extras key until you finally provide all extras, like so:

_(this is an example, no real passwords shown)_
`export AIRFLOW_CONN_SALESFORCE_DEFAULT='http://your_username:your_password@https%3A%2F%2Fyour_host.lightning.force.com?extra__salesforce__security_token=your_token&extra__salesforce__domain=&extra__salesforce__instance=&extra__salesforce__instance_url=&extra__salesforce__organization_id=&extra__salesforce__version=&extra__salesforce__proxies=&extra__salesforce__client_id=&extra__salesforce__consumer_key=&extra__salesforce__private_key_file_path=&extra__salesforce__private_key='`

### Anything else

In addition to this, the `SalesforceHook` should also accept extras without the need for the `extra__salesforce__` prefix, like many other connections do.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by alexbegg**
The `SalesforceHook` was previously requiring all connection extras, however with secrets backends it is now possible to provide one extra at a time. The current setup was directly accessing keys of the extras json which would result in a `KeyError` if you leave out one of them.

This PR allows for just the minimum required extras to be provided for each Salesforce connection method as defined in the docstring of the `SalesforceHook`: https://github.com/apache/airflow/blob/e9a72a4e95e6d23bae010ad92499cd7b06d50037/airflow/providers/salesforce/hooks/salesforce.py#L55-L60

In addition to allowing for only the required extras it also still allows for all of the extras to be provided, such as the case currently when the connection is added via the UI.

When making the connection the hook will explicitly default to `None` for all extras except for "version" (which defaults to the default API version). The reason for this is because `simple-salesforce` already has built-in authentication-choosing methods that relies on which arguments are `None` and without having `or None` in the code then setting this connection in the UI will result in the blank extras being empty strings instead of None, which would break the connection if `get` was used on its own.

The tests for the hook are also updated to only test for the required connection params when testing connections and also test that if all extras are empty strings they will be defaulted to `None`.

Also, a thank you @gigatexal, @dstandish, @josh-fell for prior PRs and PR discussions regarding this fix. This PRs incorporates the discussed points.

closes: #19506
completes PR #18929 by adding tests

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/providers/salesforce/hooks/salesforce.py
Comment (dstandish) [inline airflow/providers/salesforce/hooks/salesforce.py]: question... why are we [re?]introducing the `security_token` alongside the `extra__salesforce__`-formatted one?

is it for backward compat?  i'd assume if these were  renamed then it was done in a major and we need not worry about backward compat?

File: airflow/providers/salesforce/hooks/salesforce.py
Comment (dstandish) [inline airflow/providers/salesforce/hooks/salesforce.py]: using a cached_property instead is a common pattern in airflow and IMHO cleaner

File: airflow/providers/salesforce/hooks/salesforce.py
Comment (alexbegg) [reply airflow/providers/salesforce/hooks/salesforce.py]: In most cases when defining a connection using a secrets backend, such as via an environment variable, or using a provider's secrets backend, it is not typical to have to put the full "extra__salesforce__security_token" name. Even the connections documentation for the Salesforce provider (https://airflow.apache.org/docs/apache-airflow-providers-salesforce/stable/connections/salesforce.html) shows defining the connection as an environment variable like so:
`export AIRFLOW_CONN_SALESFORCE_DEFAULT='http://your_username:your_password@https%3A%2F%2Fyour_host.lightning.force.com?security_token=your_token'`
which is currently incorrect. Without this PR, you would need to provide it like so (with all of the extras defined:
`export AIRFLOW_CONN_SALESFORCE_DEFAULT='http://your_username:your_password@https%3A%2F%2Fyour_host.lightning.force.com?extra__salesforce__security_token=your_token&extra__salesforce__domain=&extra__salesforce__instance=&extra__salesforce__instance_url=&extra__salesforce__organization_id=&extra__salesforce__version=&extra__salesforce__proxies=&extra__salesforce__client_id=&extra__salesforce__consumer_key=&extra__salesforce__private_key_file_path=&extra__salesforce__private_key='`

So, this PR is to allow for `?security_token=your_token` as shown in the documentation, instead of having to do `?extra__salesforce__security_token=your_token`

I can't find another provider that requires the `extra__`-prefixed name to be used when defining the connection via environment variable

Comment (dstandish) [reply airflow/providers/salesforce/hooks/salesforce.py]: I see.

I am pretty sure gcp ones use the extra format.  @potiuk or @mik-laj  may be able to help guide us here

Comment (alexbegg) [reply airflow/providers/salesforce/hooks/salesforce.py]: Maybe I should split bringing in the non-prefixed extras into a seperate PR and not over-complicate this PR?

File: airflow/providers/salesforce/hooks/salesforce.py
Comment (alexbegg) [reply airflow/providers/salesforce/hooks/salesforce.py]: that would be better. I have now switched to a cached_property for self.conn

File: airflow/providers/salesforce/hooks/salesforce.py
Comment (dstandish) [reply airflow/providers/salesforce/hooks/salesforce.py]: that would probably be best.  i think it would make this PR uncontroversial.

Comment (dstandish) [reply airflow/providers/salesforce/hooks/salesforce.py]: also, looking closer at your comment, i'm not opposed to the part of this PR that makes all of the extras optional from secrets backend -- that part we should 100% do.  The part that i think would be inconsistent with past practice is to allow two different keywords for the exact same parameter.

So to use your example, what we I think we should be able to do when this pr is merged is this:
```
AIRFLOW_CONN_SALESFORCE_DEFAULT='http://your_username:your_password@https%3A%2F%2Fyour_host.lightning.force.com?extra__salesforce__security_token=your_token'
```
so you only have to specify the extra you need, _but_ you must use the same keyword as that generated by the UI.

Comment (dstandish) [reply airflow/providers/salesforce/hooks/salesforce.py]: looking more closely again:

> In most cases when defining a connection using a secrets backend, such as via an environment variable, or using a provider's secrets backend, it is not typical to have to put the full "extra__salesforce__security_token" name. 

can you provide an example?

i am only aware of GCP and it's not like this.

> Even the connections documentation for the Salesforce provider (https://airflow.apache.org/docs/apache-airflow-providers-salesforce/stable/connections/salesforce.html) shows defining the connection as an environment

You are correct it appears they missed updating the example URI when they added the UI widgets and removed `security_token` as an extra.  The doc is not correct.

Comment (dstandish) [reply airflow/providers/salesforce/hooks/salesforce.py]: ok so i did a search of the repo and found 3 other instances where both long name and short name are accepted so i withdraw my objections and let's go with it.

Comment (alexbegg) [reply airflow/providers/salesforce/hooks/salesforce.py]: I'm not considering this resolved as I would like to move the shorter names to a separate PR. Once I took it out of this PR I will then mark this resolved

File: airflow/providers/salesforce/hooks/salesforce.py
Comment (dstandish) [inline airflow/providers/salesforce/hooks/salesforce.py:34]:
```suggestion
from airflow.compat.functools import cached_property
```

File: airflow/providers/salesforce/hooks/salesforce.py
Comment (alexbegg) [reply airflow/providers/salesforce/hooks/salesforce.py]: Ok. I updated this to only include the long extra names. I am going to wait until this gets merged in before making a new PR to adding in the shorter names.

File: tests/providers/salesforce/hooks/test_salesforce.py
Comment (dstandish) [inline tests/providers/salesforce/hooks/test_salesforce.py]: this looks like it might need to be updated after the last update removing the short param names

File: tests/providers/salesforce/hooks/test_salesforce.py
Comment (dstandish) [inline tests/providers/salesforce/hooks/test_salesforce.py]: I see that you have added parameterization in 4 places, removing the "empty string" extras.

I think we can simplify this and be more direct about what we're testing.

What do you think about the following.

Add one test that verifies that when extra params are supplied as empty string, the call converts those params to `None`

Then, for all of the tests you have parameterized, instead of parameterizing, just chop down the params to only the non-empty-string params.  We'll know they will be converted to `None` since you will add the test suggested above, so they no longer need to be included in those 4 tests.  This leaves just your "required params" sets.

**Review by dstandish — CHANGES_REQUESTED** [review]

File: airflow/providers/salesforce/hooks/salesforce.py
Comment (dstandish) [reply airflow/providers/salesforce/hooks/salesforce.py]: Sounds good. Just reviewed

File: tests/providers/salesforce/hooks/test_salesforce.py
Comment (alexbegg) [reply tests/providers/salesforce/hooks/test_salesforce.py]: Correct, I missed this

File: tests/providers/salesforce/hooks/test_salesforce.py
Comment (alexbegg) [reply tests/providers/salesforce/hooks/test_salesforce.py]: I agree. That sounds good actually. A lot less lines of code.

File: airflow/providers/salesforce/hooks/salesforce.py
Comment (alexbegg) [reply airflow/providers/salesforce/hooks/salesforce.py:34]: One of the PR's tests is failing `ModuleNotFoundError: No module named 'airflow.compat'`. Is it possible airflow.compat can't be used in provider packages? The test that is failing is "Tests / Build and test provider packages wheel (pull_request)"

Comment (dstandish) [reply airflow/providers/salesforce/hooks/salesforce.py:34]: yeah probably best to just switch that back.  i'll connect with jarek re guidance going forward.  thank you

**Review by dstandish — APPROVED** [review]
Looks good to me, pending the cached_property import fix

**Comment by alexbegg** [comment]
> Looks good to me, pending the cached_property import fix

The import is fixed

**Comment by dstandish** [comment]
excellent, thank you

**Comment by gigatexal** [comment]
Thank you @alexbegg for getting this across the line!

**Comment by halilduygulu** [comment]
thanks for the change. in which version this will be released? https://pypi.org/project/apache-airflow-providers-salesforce/#history

**Comment by potiuk** [comment]
In the next one released ~ the end of the month

**Comment by halilduygulu** [comment]
and was changing extra parameter names absolutely necessary? I have to change this 40 times in our airflow setup, not everyone is creating connections from ui.
```
security_token -> "extra__salesforce__security_token" 
domain -> "extra__salesforce__domain"
```

**Comment by dstandish** [comment]
I totally feel your pain.  I do not like the decision.  My very uninformed guess is that more often than not creds are not managed in UI and I would prefer if we never implemented the UI customizations.  I think @alexbegg may also have been working on a PR to make it so that you can optionally use the "normal" names instead.  But clearly that won't help you in time.

BUT you have an option.  Since all the conn extra parsing is done in `get_conn` you could just subclass the hook and override the  behavior so that it looks for the normal names.  Then if you are using your own salesforce operator you can just point to your salesforce hook.  And if your not you could create a subclass of the salesforce operator that uses _your_ salesforce hook.  Depending on your setup this could be an easier path for you.

**Comment by alexbegg** [comment]
Yes, I am planning to make a separate PR to have `get_conn` to allow both formats, `security_token` and `extra__salesforce__security_token`. This PR initially allowed for that however I figured it was too much discussion regarding that to be incorporated into this PR. There is a discussion on the dev list about why the names changed.

**Comment by alexbegg** [comment]
This PR is using the same extra names as the previous version of the method. Once I make the PR I'll link it here. Also I'm a bit busy at the moment if someone want to handle making a PR in my place for that I'll appreciate that!

**Pull request merged**
